### ISSUE TITLE
Preserving Item Mod Data Megathread

### ISSUE THREAD

**Issue by JavidPack** [tModLoader/tModLoader#265]
This issue serves as a home for all issues pertaining to preserving mod data in Mod Items through all possible movement vectors. 

All items are tested with ExampleMod's ExampleInstancedGlobalItem.

### Single Player
- [ ] Inventory
  - [x] Save/Load
  - [ ] Quick stack to chests (see #235)
  - [x] Linking an item into chat (by holding LALT)
  - [x] Preserve through Unload (Data Lost as of Nov 11, 2017)
    - [x] Global Item data recovered from MysteryGlobalItem
    - [x] MysteryGlobalItem preserved through reloads
- [x] Chest
  - [x] Save/Load in Chest
  - [x] Preserve through Unload (Data Lost as of Nov 11, 2017)
- [x] Shop Sell/Buy back
- [x] Throw into world
- [x] Reforge ?
- [x] Tooltip (Clone)
- [x] Shift Click to Chest/Trash/Shop
- [ ] Right Click to/from Equipment ?
- [x] Tile Entities and Special Tiles
  - [x] Item Frames #207 

### 1.4 Update
- [ ] Void Storage
- [ ] ItemDropRules

### Multiplayer
TODO: Untested
- [ ] Quick Stack #235 

## ModItem
TODO: Untested

**Comment by Chicken-Bones** [comment]
If you want to handle the testing for this, I'm quite happy to do the fixing and verify everything is going through the appropriate channels (since I did the Save/Load, Networking and advised on Mod/Global design)

Also are we talking about ModItem data or GlobalItem data in each case, and as a sub point for Save/Load/Unload, are we talking about when the mod providing the GlobalItem is unloaded, or when the mod providing the host item with GlobalItem data from another mod is unloaded?

**Comment by JavidPack** [comment]
For the preserve through unload, what I tried so far was a vanilla item with GlobalItem data attached (ExampleInstancedGlobalItem, it attaches a string to the item). The mod supplying the GlobalItem was unloaded, the world entered and exited, the mod was re-enabled, and the world was re-entered. In this case, the data was still present in the .twld file after leaving the world while the mod was unloaded, but the MysteryGlobalItem was not restored to the Player or World once the mod was reenabled. Also, if the world is entered while the data is already in MysteryGlobalItem, the data is lost as well. (As in, Exit World/Unload/Enter/Exit/Enter/Exit)

**Comment by Jofairden** [comment]
#257 should fix for reforges. ( I did test, worked for me )

**Comment by Jofairden** [comment]
I added `Linking an item into chat (by holding LALT)` as it also suffers from this issue

**Comment by Jofairden** [comment]
I believe `Preserve through Unload` is fixed by #329

**Comment by Jofairden** [comment]
To add a comment on quick stacking: As discussed before, stacking is undefined behavior with custom data. To support this usecase, we need hooks that can define behavior when stacking (CanStack, OnStack.. etc. as discussed). This is probably best suited as a separate issue at some point.

**Comment by Jofairden** [comment]
Adding a couple new sources to the list for 1.4:
- Void Storage
- Item drop system (Drop loot rules)

**Comment by Solxanich** [comment]
So, I'm just trying to figure out what's left to tackle here. Allow me to summarize knowns

Tile entities, item frames were addressed in Dec by Mirsario: [12cd989](https://github.com/tModLoader/tModLoader/commit/12cd989323211a3355de77f7ac234d188738ea09)

Mannequins preserving armor was also addressed at some point, though can't find any particular source (is this what right-click to equipment refers to? Or are we just speaking about the right click to equip, at which point that's been working afaik?)

Preservation of the items in chests will be addressed by: #1266 

Preservation of items in void vault will be addressed by: #1270 

NPC Loot drop system was addressed by: #1063 
-------------------------------------------------------------------------
tldr; is all that is left fixing the quick stack for chests in multi?

**Comment by Solxanich** [comment]
Looks like quick stack was fixed at some point for multi, or at least compared original issue #235 and never marked here as done (proof with reported durability mod):

https://user-images.githubusercontent.com/59670736/104511163-8f7b7980-55a9-11eb-8eff-dfc664446aee.mp4

So I guess this is almost closable?

**Comment by JavidPack** [comment]
This needs a once over test, then probably close in favor of individual issues still present logged in existing issues.

**Comment by Chicken-Bones** [comment]
I believe this is done
